### ISSUE TITLE
Support Backblaze B2

### ISSUE THREAD

**Issue by sdbarker** [duplicati/duplicati#1498]
I'd love to have a backend for the new B2 offering from Backblaze.

https://www.backblaze.com/b2/cloud-storage.html

**Comment by kenkendk** [comment]
I have signed up for beta access, the API looks simple enough

**Comment by sdbarker** [comment]
Thanks Kenneth.

**Comment by sdbarker** [comment]
Awesome, thanks for getting this added!  I'm having an issue though, when adding the bucket, both by name and by ID, Duplicati says that the folder doesn't exist and asks to create it.  If I tell Duplicati to create the bucket, it returns a 400, bad bucket name, and that the bucket must be alphanumeric and '-'.  The name meets those standards (and already exists).

**Comment by kenkendk** [comment]
Do you get the error from the commandline or the gui ?

**Comment by sdbarker** [comment]
The error comes from the GUI.

**Comment by kenkendk** [comment]
I have updated the UI, and tested that I could create a bucket.
Can you confirm that the problem has been fixed?

**Comment by sdbarker** [comment]
Looks to be resolved Ken, thanks!

**Comment by sdbarker** [comment]
It seems as though there's a new issue.  I feel like it may be related to the authorization token expiring, but I'm now getting a 500 from the UI.  It feels like the UI is returning the 500, and I have a hunch I'm getting a 401 from B2.  It may be necessary / prudent to authorize_account before any subsequent API calls, or at least at a periodic interval so that long-running backups don't have their authorization token expire while they're running.

Here's a relevant stack trace:

```
{"ClassName":"System.Exception","Message":"500 - internal_error: Internal Error","Data":null,"InnerException":null, "HelpURL":null,"StackTraceString":"   at Duplicati.Library.Main.BackendManager.List() in [REDACTED]Duplicati\\Library\\Main\\BackendManager.cs:line 1028\r\n   at Duplicati.Library.Main.Operation.FilelistProcessor.RemoteListAnalysis(BackendManager backend, Options options, LocalDatabase database, IBackendWriter log) in [REDACTED]Duplicati\\Library\\Main\\Operation\\FilelistProcessor.cs:line 177\r\n   at Duplicati.Library.Main.Operation.FilelistProcessor.VerifyRemoteList(BackendManager backend, Options options, LocalDatabase database, IBackendWriter log) in [REDACTED]Duplicati\\Library\\Main\\Operation\\FilelistProcessor.cs:line 80\r\n   at Duplicati.Library.Main.Operation.TestHandler.Run(Int64 samples) in [REDACTED]Duplicati\\Library\\Main\\Operation\\TestHandler.cs:line 54\r\n   at Duplicati.Library.Main.Controller.RunAction[T](T result, String[]& paths, Action`1 method) in [REDACTED]Duplicati\\Library\\Main\\Controller.cs:line 355\r\n   at Duplicati.Library.Main.Controller.Test(Int64 samples) in [REDACTED]Duplicati\\Library\\Main\\Controller.cs:line 288\r\n   at Duplicati.Server.Runner.Run(IRunnerData data, Boolean fromQueue) in [REDACTED]Duplicati\\Server\\Runner.cs:line 462","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":"8\nList\nDuplicati.Library.Main, Version=2.0.0.7, Culture=neutral, PublicKeyToken=8bfe994a39631a7b\nDuplicati.Library.Main.BackendManager\nSystem.Collections.Generic.IList`1[Duplicati.Library.Interface.IFileEntry] List()","HResult":-2146233088,"Source":"Duplicati.Library.Main","WatsonBuckets":null}
```

**Comment by kenkendk** [comment]
The authorization is repeated every hour, so that long running backups should not have any issues.
I can reduce that time, but there is no documentation for how long it is supposed to be valid:
https://www.backblaze.com/b2/docs/b2_authorize_account.html

The "500 - internal_error" is from the Backblaze server, there is no such string in Duplicati.

**Comment by sdbarker** [comment]
It doesn't look like there's a logging level that dumps out the exact response that's received, is that correct?  That would be helpful to debug issues like this.

**Comment by kenkendk** [comment]
Unfortunately, that is all the information returned by the server:
`500 - internal_error: Internal Error`

The stack trace indicates that the error occurs when attempting to list the remote folder.

If I get a clue as to what happens in your setup that does not happen in mine, I will try to reproduce.

1) Can you log on to the Backblaze page and see your data?
2) How many files do you have?
3) How many GB does the data fill up?
4) Does the error occur with all operations?
5) Did it start occuring after something particular?

**Comment by sdbarker** [comment]
Kenneth,

According their team, the docs state:

```
Failure codes in the range 500 through 599 mean that the storage pod is having trouble accepting your data. In this case you must call b2_get_upload_url to get a new uploadUrl and a new authorizationToken. The reason for this is that the individual storage pod you are uploading into may be full of data and thus will never accept any more data for now, or it has crashed, or has been placed into maintenance and will not accept data, etc. No matter what the reason, you must use b2_get_upload_url to get a new uploadUrl and new authorizationToken in order to retry the upload.
```

So it looks like on any 500, we'll need to call that other endpoint before retrying an upload.

To answer your questions:

1.  Yes
2.  Somewhere between 1,000 and 2,000 when this happens.
3.  Somewhere between 40 and 80gb
4.  It seems to only be on uploads.
5.  I haven't been able to find a reliable repro -- the closest that I get is configure a new backup job using B2 as the backend, have a large set of data that's being processed, and run the job.

With all of that said, I think the answer lies with the information provided by Backblaze (I've actually pointed them at this ticket).

**Comment by kenkendk** [comment]
That happens as it is.

If the operation fails (for any reason, including a 5xx status), the backend instance is disposed, and a new instance is created.
When that happens, a new instance will re-authorize and get a new auth-id.

The init code for the backend is here:
https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Backend/Backblaze/B2.cs#L90
The initialization of the uploadurl (and token) is here:
https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Backend/Backblaze/B2AuthHelper.cs#L61
The exception handing is here:
https://github.com/duplicati/duplicati/blob/master/Duplicati/Library/Main/BackendManager.cs#L450

I am guessing something else is wrong. I will see if I can run a test with 2000 files being generated tomorrow.

Btw, you say "only uploads", but the trace you provided seems to happen for a list operation?

**Comment by sdbarker** [comment]
Yeah, the more time I spent with the code, it looked like things were being handled in accordance with the API.  I'll see if I can spend more time with it, and please let me know what you find from your larger test.

I did say only uploads and did provide a trace pointing to a list op, you're right.  I don't see anything about the 500-599 status codes in the API docs for b2_list_file_versions, but I'm wondering if that 500-599 range is thrown for any call when the current pod is no longer valid.

**Comment by nilayp** [comment]
Nilay from Backblaze here.

While technically any API can throw a 5xx, this would be a bug or operational problem with B2. The only exception is the b2_upload_file API. This API will return a 5xx error if the endpoint goes down and the developer is expected to handle the issue by fetching a new upload URL (and token) using b2_get_upload_url and then using these to re-upload the file that failed.

One pro tip: When uploading n files, you can call b2_get_upload_url n times and upload all these files in parallel. Of course, you should calibrate n correctly so that you don't saturate I/O on your computer or the network connection. Spinning off multiple upload (and download too!) threads is the key to fast performance.

When you feel the integration is ready - let me know. I'll add it to the integrations page on our website.

-- Nilay

**Comment by kenkendk** [comment]
@nilayp Thanks for commenting!
As I wrote above, the re-upload is handled as it should be.

I can now reproduce the problem with 4000 files in a bucket.
All uploads work as expected, no errors for any of the 4000 uploads.
But... when listing the bucket contents, using `b2_list_file_versions` I get repeated 500 - internal error.

Digging a little deeper, I can see that the initial request works nicely, but the next (paged) request fails. I have a page size of 200, and get a valid-looking `nextFileId`, but when I call `b2_list_file_versions` it gives the 500 error.

**Comment by kenkendk** [comment]
And I just discovered the reason for that... 

I had initially implemented the backend using `b2_list_file_names` and was using `startFileName` for paging. When I changed to `b2_list_file_versions`, I switched to using `startFileId`, but I must have forgotten to test the paging. Re-reading the documentation, I can see that a "fileid" is not a globally unique identifier, so the paging does not work unless you specify both `startFileId` and `startFileName`.

**Comment by sdbarker** [comment]
@kenkendk I've got a backup going right now in the debugger to see if I hit a couple of breakpoints trying to trace the issue.  Should I stop the backup and wait to pull a new commit from you?

**Comment by kenkendk** [comment]
Your call, I put up the commit

**Comment by sdbarker** [comment]
Sounds good.  I've got it pulled down.  After my bucket empties, I'll run
it and let you know how it goes.


On Tue, Dec 15, 2015 at 12:10 PM, Kenneth Skovhede <notifications@github.com

**Comment by kenkendk** [comment]
I tested it again today, and it ran to completion with 4000 files (well... it stopped due to exceeding the download quota).

**Comment by nilayp** [comment]
Kenneth - can you please release a version with the fix above? We have some 100k plus file test cases I can put it through here.
What I don't have is a windows developer environment to build from source... :(

Also - would love to chat about communicating Duplicati to our windows server backup customers. Can you drop me an email at b2contact@backblaze.com? Just put my name in the message and the support team will escalate the ticket to me.

-- Nilay

**Comment by kenkendk** [comment]
@nilayp: You can build it (and run it) on any OS, not just Windows. I use OSX for instance.
Anyway, there is a build here: http://updates.duplicati.com/preview/duplicati-2.0.0.96_preview_2015-12-17.zip

I will write you a mail.

**Comment by thejargen** [comment]
Any chance of a linux build of that? Having no luck compiling it myself.

**Comment by robsonsobral** [comment]
Hi!

I'm testing here, and, for a bucket called `duplicati-test`, it always try to create a folder with the same name, it doesn't matter if a set a folder or not:

```
The folder duplicati‑test does not exist Create it now?
```

If I choose yes:

```
Failed to connect: 400 - bad_json: Invalid characters in bucketName: must be alphanumeric or '-'
```

Firefox 46 on Windows 10.

Thanks, guys!

**Comment by robsonsobral** [comment]
Update.

I tested again today, twice, and got no errors.

**Comment by kenkendk** [comment]
So this is fixed?

**Comment by robsonsobral** [comment]
Apparently. Maybe it was an error on their side? I don't know, but it's working now.

**Comment by kenkendk** [comment]
I fixed an additional bug today, with spaces in the path (spaces in the filenames were handled already) 0e8276201f2ab99dc4f580fc0c51c4f3b744a188.
